DEP: deprecate scalar conversions for arrays with ndim > 0

nschloe opened the pull request:
This PR reflects some of the progress achieved in issue #10404 and is used to asses the impact of the changes. 

With the changes in this PR, `float(numpy.array([1.0])` now gives a warning; likewise some other things:
```python
import numpy

a = numpy.random.rand(10, 1)
a[0] = numpy.array([1.0])       # okay
a[0] = numpy.array(1.0)         # okay
a[0] = 1.0                      # okay

b = numpy.random.rand(10)
b[0] = numpy.array([1.0])       # ValueError: setting an array element with a sequence.
b[0, ...] = numpy.array([1.0])  # okay
b[0] = numpy.array(1.0)         # okay
b[0] = 1.0                      # okay
```
This aligns the behavior of numpy arrays with that of lists:
```python
float([3.14]) 
```
```
TypeError: float() argument must be a string or a number, not 'list'
```
```python
import numpy as np

a = np.random.rand(5)
a[0] = [3.14]
```
```
ValueError: setting an array element with a sequence.
```
Fixes #10404.

eric-wieser commented on numpy/core/function_base.py:
```diff
@@ -142,7 +142,7 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None):
 
     if endpoint and num > 1:
-        y[-1] = stop
+        y[-1,...] = stop
```
Why is this needed? Seems the problem is in the callers.

nschloe replied:
If `stop` is a scalar, this works fine without `...`. However, linspace also accepts `Arrayish` inputs for `start` and `stop`. See https://github.com/numpy/numpy/issues/3504.

eric-wieser commented:
Would be interesting to see how extensive the changes need to be to make CI pass, and whether this impacts the scipy or pandas test suites.

eric-wieser commented:
Looks like randomgen added more tests that rely on this behavior, should be straightforward to just remove them

nschloe commented:
I'm on it.

eric-wieser commented on numpy/random/_bounded_integers.pyx.in:
```diff
@@ -314,10 +314,15 @@ cdef object _rand_{{nptype}}(object low, object high, object size,
+    # nschloe, 2019-11-29:
+    # This conditional allows x and [x]. Why stop there? [[[[[x]]]]] also wants to be a
+    # low/high. It'd be more consistent if we just allowed int input for low, high in
+    # the first place, not array_like.
```
I'd be inclined to just change the test here, rather than trying to allow an array

nschloe replied:
The documentation says: `low : int or array-like` (likewise `high`). Should I change that?

eric-wieser replied:
Yes

eric-wieser replied:
Perhaps remove this weirdness in a separate pr?

eric-wieser replied:
Ah, hang on - this is just an optimisation. There's a full broadcast path below. I'd just remove this path for all but the 0d case

nschloe replied:
Right now, it's possible to do `randint([0, 1], [4, 7])` which gives two ints, one between 0 and 4 and another between 1 and 7. This has been introduced sometime after 1.16. I've personally never used this but I'm not sure if this is a common use case. We could allow in arrays of arbitrary shape and have the output of shape `low_shape + size` (`low_shape == high_shape` of course).

nschloe replied:
> I'd just remove this path for all but the 0d case

Indeed.

eric-wieser commented on numpy/testing/_private/utils.py:
```diff
@@ -660,7 +660,6 @@ def assert_approx_equal(actual,desired,significant=7,err_msg='',verbose=True):
     __tracebackhide__ = True  # Hide traceback for py.test
     import numpy as np
 
-    (actual, desired) = map(float, (actual, desired))
```
Again, I think this is actually correct, and it's the caller that is wrong

nschloe commented:
So I've been working on scipy a bit (https://github.com/scipy/scipy/pull/11147) and found that in parts it was quite lax when it came to float conversion of arrays. I'll fix all the inconsistencies there and in panda (if any).

All of what I discovered are actual bugs which is what makes me like this PR here even more. It _will_ break (buggy/inefficient) code though, so there'll perhaps have to be a deprecation process. 

What do you think?

eric-wieser commented:
To me there's no question that if we do this, it will need to be after a deprecation period. I'm definitely in favor of tracking down bugs in downstream projects with this change, whether or not we decide to merge this. Make sure to link those fixes to here so we can get a feel for the impact.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2577,8 +2577,9 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+
+    if (PyArray_NDIM(self) != 0) {
+        PyErr_SetString(PyExc_TypeError, "only rank-0 arrays can "\
                         "be converted to Python scalars");
```
Responding to @seberg's question in the corresponding issue - this, and the other instance, should be:
```
if (PyArray_NDIM(self) != 0) {
    if (PyArray_SIZE(self) == 1) {
        if (DEPRECATE("Conversion of an array with one element to a scalar is deprecated, and will error in future") < 0) {
            return NULL;
        }
    }
    else {
        PyErr_SetString(PyExc_TypeError, "only rank-0 arrays can "
                        "be converted to Python scalars");
        return NULL;
    }
}
```

seberg replied:
Yeah, when I read the stuff, I thought of `PyArray_Result` first, thats why I thought it would be super complicated. This is fine, it is mostly about how many downstream pain credits we can/should spend on it.

eric-wieser replied:
Probably worthwhile extracting a helper function so this code isn't duplicated.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2582,8 +2582,8 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
-                is deprecated, and will error in future. \
-                Ensure you extract a single element from your array before performing this operation"
+is deprecated, and will error in future. \
+Ensure you extract a single element from your array before performing this operation"
```
You're looking for
```C
            if (DEPRECATE(
                "Conversion of an array with one element to a scalar "
                "is deprecated, and will error in future. "
                "Ensure you extract a single element from your array before performing this operation."
            ) < 0) {
```

eric-wieser commented on numpy/core/src/multiarray/common.c:
```diff
@@ -945,5 +945,24 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
 
-
-
+bool
```
```suggestion
NPY_NO_EXPORT bool
```

eric-wieser commented on numpy/core/src/multiarray/common.c:
```diff
@@ -945,5 +945,24 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
-
-
+bool
+has_rank_0(PyArrayObject *v)
```
Probably better to call this
```suggestion
check_has_rank_0(PyArrayObject *v)
```
and give it a `int` return of either `0` or `-1`, to make it clear that it sets an error

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -7578,9 +7582,10 @@ def test_to_int_scalar(self):
+            assert_equal(int_func(np.array(1)), 1)
+            assert_equal(int_func(np.array(0)), 0)
+            assert_raises(DeprecationWarning, int_func, np.array([1]))
+            assert_raises(DeprecationWarning, int_func, np.array([[42]]))
```
This should be:
```
with pytest.assert_warns(DeprecationWarning):
    assert_equal(int_func(np.array([[42]])), 42)
```
or similar

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -3299,10 +3299,8 @@ def test__complex__(self):
-            cp = complex(c)
-            assert_equal(cp, c, msg)
+            assert_warns(DeprecationWarning, complex, b)
+            assert_warns(DeprecationWarning, complex, c)
```
You should be able to simultaneously assert that this warns and check its value here

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -7578,9 +7582,11 @@ def test_to_int_scalar(self):
-            assert_equal(int_func(np.array([[42]])), 42)
+            assert_equal(int_func(np.array(1)), 1)
+            assert_equal(int_func(np.array(0)), 0)
+            with assert_warns(DeprecationWarning):
```
You need a with block for each function call, so that you can prove that _both_ warn

nschloe commented:
pandas builds and tests cleanly against this PR.

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2577,10 +2577,9 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
-        return NULL;
+
+    if (check_has_rank_0(self) < 0) {
+      return NULL;
```
```suggestion
        return NULL;
```

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -3323,9 +3328,15 @@ def test__complex__should_not_work(self):
+        # (because of the []). Not sure how to deal with it.
+        # e = np.array(['1+1j'], 'U')
+        # assert_raises(TypeError, complex, e)
+        # assert_warns(DeprecationWarning, complex, e)
```
I think you can use `assert_warns` wrapping the original `assert_raises`

eric-wieser commented on numpy/random/_bounded_integers.pyx.in:
```diff
@@ -314,8 +314,7 @@ cdef object _rand_{{nptype}}(object low, object high, object size,
     high_ndim = np.PyArray_NDIM(high_arr)
-    if ((low_ndim == 0 or (low_ndim == 1 and low_arr.size == 1 and size is not None)) and
-            (high_ndim == 0 or (high_ndim == 1 and high_arr.size == 1 and size is not None))):
+    if (low_ndim == 0 and high_ndim == 0):
```
nit:
```suggestion
    if low_ndim == 0 and high_ndim == 0:
```

eric-wieser replied:
Want to spin out a PR with this fixup in to be merged immediately?

eric-wieser commented on numpy/testing/tests/test_utils.py:
```diff
@@ -609,8 +609,8 @@ def setup(self):
         self._assert_func = assert_approx_equal
 
     def test_simple_arrays(self):
```
Since this was never intended to be used on 1d arrays:
```suggestion
     def test_simple_0d_arrays(self):
```

eric-wieser commented on numpy/testing/tests/test_utils.py:
```diff
@@ -609,8 +609,8 @@ def setup(self):
-        x = np.array([1234.22])
-        y = np.array([1234.23])
+        x = np.array(1234.22)
+        y = np.array(1234.23)
```
Want to spin out a PR with this fixup in to be merged immediately?

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -7578,9 +7589,12 @@ def test_to_int_scalar(self):
-            assert_equal(int_func(np.array([0])), 0)
-            assert_equal(int_func(np.array([[42]])), 42)
+            assert_equal(int_func(np.array(1)), 1)
+            assert_equal(int_func(np.array(0)), 0)
```
These two lines could go into the spin-off PR too

eric-wieser commented on numpy/core/src/multiarray/common.c:
```diff
@@ -945,5 +945,24 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
+        }
+    }
+    return 0;
+}
```
super nit: could eliminate some nesting with
```C
check_has_rank_0(PyArrayObject *v)
{
    if (PyArray_NDIM(v) == 0) {
        return 0;
    }
    /* Remove this if when the deprecation expires */
    if (PyArray_SIZE(v) == 1) {
        /* Numpy 1.19.0, 2019-12-18 */
        if (DEPRECATE(
            "Conversion of an array with one element to a scalar "
            "is deprecated, and will error in future. "
            "Ensure you extract a single element from your array before performing this operation."
        ) < 0) {
            return -1;
        }
        return 0;
    }
    PyErr_SetString(PyExc_TypeError, "only rank-0 arrays can "
                        "be converted to Python scalars");
    return -1;
}
```

eric-wieser commented:
One failure remains:

https://dev.azure.com/numpy/numpy/_build/results?buildId=6980&view=ms.vss-test-web.build-test-results-tab&runId=1082926&resultId=101493&paneView=debug

nschloe commented:
Added to #15126.

nschloe commented:
Right now looking at pandas. (I reported earlier that it builds cleanly, but that's not the case -- You have to run the tests with `-Werror`.)

eric-wieser commented:
It's reassuring how few changes were required in #15126 and our test suite.

nschloe commented:
matplotlib requires one change for the tests to pass without warning. https://github.com/matplotlib/matplotlib/pull/16003

seberg commented:
It sounds like we probably won't be pushing this forward right now (unfortunately). The issue is that it seems like probably have to spend a lot of "breaks user code" chips on it, and not everyone is even quite convinced that it is incorrect, as opposed to just slightly weird semantics. Sorry about this, but it sounds like the chance is that we will be closing this soon for now :/.

My opinion: Once all the major packages have been fixed for a while, and with a few examples of how this creates actual bugs, we may be able to reconsider (I would actually like that). But I do not have examples of bugs and write it up for the mailing list. If we can convince that deprecating this (very very slowly) fixes some bugs and is likely not to annoy too many users out there, there is a case that can be made.

charris commented:
@seberg This should be closed?

seberg commented:
I am not sure. I still think it is desirable, but whether or not we should go ahead depends a bit on how the major packages would break with it. @nschloe do you remember the status?

nschloe commented:
> I am not sure. I still think it is desirable, but whether or not we should go ahead depends a bit on how the major packages would break with it. nschloe do you remember the status?

At the the time, I probed some of the biggest NumPy-dependent Python packages (scipy, sympy, mpl, astropy I think) and submitted a bunch of PRs. Not sure of the status there, but I can check again.

I also think this is still desirable, so perhaps this can be left open as an "epic" of sorts.

eric-wieser commented:
I remember large chunks of scipy that were sloppy about this, but I think a motivated user could fix them one by one. So I'd agree with leaving this open, in case anyone with such motivation appears.

nschloe commented:
- [ ] matplotlib https://github.com/matplotlib/matplotlib/pull/19952

nschloe commented:
Whenever I get back to scipy, open an issue or even worse, a PR, it makes me appreciate the numpy community so much more. So thanks guys for sticking around. :+1:

nschloe commented:
> I remember large chunks of scipy that were sloppy about this, but I think a motivated user could fix them one by one. 

I'm happy to report that said motivated individual has plowed through the scipy issues and fixed all of them in 3 PRs. (See above.) Most of the issues were due to a sloppy handling of goal functionals not returning actual scalars. All the changes have improved the scipy code.

I had already fixed matplotlib and pandas a while ago; the changes there where miniscule.

I would consider this PR ready for merge and would be happy about a re-review.

mattip commented:
This still needs a release note

eric-wieser commented on numpy/core/src/multiarray/common.c:
```diff
@@ -476,3 +476,25 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
+    PyErr_SetString(PyExc_TypeError, "only rank-0 arrays can "
+                        "be converted to Python scalars");
+    return -1;
+}
```
```suggestion
}

```

eric-wieser commented on numpy/core/src/multiarray/common.h:
```diff
@@ -345,3 +345,5 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
 #endif
 
+NPY_NO_EXPORT int
+check_has_rank_0(PyArrayObject *v);
```
```suggestion
check_has_rank_0(PyArrayObject *v);

```

nschloe commented:
@mattip Release note added.

h-vetinari commented on numpy/core/src/multiarray/common.c:
```diff
@@ -476,3 +476,25 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
+    }
+    /* Remove this if when the deprecation expires */
+    if (PyArray_SIZE(v) == 1) {
+        /* Numpy 1.19.0, 2019-12-18 */
```
This line likely needs updating too...

nschloe replied:
Thanks, fixed now.

h-vetinari commented on numpy/core/src/multiarray/common.c:
```diff
@@ -484,7 +484,7 @@ check_has_rank_0(PyArrayObject *v)
     /* Remove this if when the deprecation expires */
     if (PyArray_SIZE(v) == 1) {
-        /* Numpy 1.19.0, 2019-12-18 */
+        /* Numpy 1.21.3, 2021-08-17 */
```
Not to be too picky, but I presume this is not going to be backported, so should probably say `1.22.0`.

nschloe replied:
Ah yes. Should the deprecation note go in `doc/source/release/1.22.0-notes.rst`, too? (That file is still more or less empty.)

seberg replied:
It should go as a deprecation into `doc/release/upcoming_changes`, check the `README` in that file.  The release note is incorrect as is.

seberg reviewed (commented):
It would be good to have a test in `core/tests/test_deprecations.py`.  It is not super important to be in that file, but it keeps deprecations in one place and also ensures that we test the error path.

(I don't see code coverage complaining, although I am not sure what would be covering the error paths).

nschloe commented on numpy/core/src/multiarray/common.c:
```diff
@@ -476,3 +476,25 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
+            "Conversion of an array with ndim > 0 to a scalar "
+            "is deprecated, and will error in future. "
+            "Ensure you extract a single element from your array before performing this operation."
+        ) < 0) {
```
Fixed.

nschloe commented:
> It would be good to have a test in core/tests/test_deprecations.py. It is not super important to be in that file, but it keeps deprecations in one place and also ensures that we test the error path.
> (I don't see code coverage complaining, although I am not sure what would be covering the error paths).

I added some explicit deprecation tests now. The previous coverage came from some `assert_warns(DeprecationWarning)` in `numpy/core/tests/test_multiarray.py` (see diff).

nschloe commented:
If there's anything I can do to advance this, let me know.

seberg commented:
While I personally like this, it might also be a bit of my bias towards consistency vs. some (rare/non rare?) convenience.  Could you hit the mailing list outlining the proposal and what it improves and breaks?

I tend towards trying it and roll-back if there is push-back, but I am not sure it is important enough to fight hard for it.

nschloe commented:
@asmeurer came up with an interesting example. It goes something like
```python
import numpy as np

a = np.random.rand(2, 3)
a[0] = [1, 2, 3]                # ok
a[0] = np.array([1, 2, 3])      # ok
# a[0] = [[[[1,2,3]]]]          # ValueError: setting an array element with a sequence.
a[0] = np.array([[[[1,2,3]]]])  # ok?!
```
The last line does something like a squeeze before setting the array, and is not intercepted by the changes in this PR. It can be thought of an exension of the
```python
a = np.random(3)
a[0] = np.array([1])
```
that we'd like to deprecate. I'm not sure how common assignments like `a[0] = np.array([[[[1,2,3]]]])` are, but I'm leaning towards deprecating it too for the sake of consistency with lists and consistency with the deprecation in this PR. Not sure if it should be done as part of this PR. What do y'all think?

nschloe commented:
> I was aware of it, and always considered this part of the deprecation here. I am not sure why you see it as different from your example of arr1d[0] = [1.]?

The only difference is that
```python
import numpy as np

a = np.random.rand(2, 3)
a[0] = np.array([[[[1,2,3]]]])
```
does _not_ yield a warning even after this PR. Question: Should it be added in another PR, or in this?

seberg commented:
Oh, I thought this PR would affect that as well!  Maybe the reason why it was so painless;  so there may be a point in not including it just to make it a smaller, less painful, step...

nschloe commented:
Nah, just scalars for now. I think that's already an improvement over the current behavior. I think it'd make sense to have the multidim case follow in a second PR.

Micky774 commented:
@nschloe Are you still interested in continuing this PR? If so, I have a couple modifications that reconcile changes w/ upstream main and fix the failing tests that I can open a PR to your branch with. If you're not interested in / able to continue this PR, I wouldn't mind picking it up :)

nschloe commented:
I sure am! Not sure what exactly needs to be done though. Perhaps another run on SciPy, Pandas, etc. to see if anything has crept up in the meantime.

Micky774 commented:
> I sure am! Not sure what exactly needs to be done though. Perhaps another run on SciPy, Pandas, etc. to see if anything has crept up in the meantime.

Hey there! I hadn't heard from you in a while so I went ahead and opened a follow-up PR #21846 to ensure that the great work you did would make it through to merge. I preserved your commits so you retain credit in the eventual squash-merge authorship. Not sure how the maintainers want to reconcile this, but feel free to hop over there and offer some feedback on the modifications I made as well :)

nschloe commented:
> Hey there! I hadn't heard from you in a while so I went ahead and opened a follow-up PR

You opened that about 12 hours after you asked here.

Micky774 commented:
@nschloe I'm happy to close my PR. The reason I opened my PR so quickly after asking was because I saw that there was no activity on this PR for ~7 months already, and because I wouldn't mind closing it if you were interested in continuing this one. 

I just want to make it clear that this wasn't in interest of claiming credit for your work -- in the main body of my PR I laid out explicitly that the significant contributions were yours and that mine were minor reconciliations. Sorry if this caused you to worry that the work you did was going to be undermined. Again, it really is good work and I apologize that my actions caused an unnecessary stir.

mattip commented:
There are some failing tests.

mattip commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2761,9 +2761,7 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
-    if (PyArray_SIZE(self) != 1) {
-        PyErr_SetString(PyExc_TypeError,
-                "only length-1 arrays can be converted to Python scalars");
+    if (check_has_rank_0(self) < 0) {
```
This needs to be declared in a header. I wonder how it passes compilation?

nschloe commented:
So this
```python
arr = np.ones((1, 1))
np.array([0, arr], dtype=np.float64)
```
Results in _two_ deprecation warnings after this PR. One is old,
```
<ipython-input-3-4e22c1d93d88>:1: DeprecationWarning: setting an array element with a sequence. This was supported in some cases where the elements are arrays with a single element. For example `np.array([1, np.array([2])], dtype=int)`. In the future this will raise the same ValueError as `np.array([1, [2]], dtype=int)`.
  np.array([0, arr], dtype=np.float64)
```
and one is new,
```
<ipython-input-3-4e22c1d93d88>:1: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.24.)
  np.array([0, arr], dtype=np.float64)
Out[3]: array([0., 1.])
```
The latter occurs because internally the `arr` is converted to a scalar. It's nice to see the deprecations aligning, but we probably don't want two.

seberg commented:
Lets not worry about the duplicate.  In fact, I think I may just finalize that deprecation now.

nschloe commented:
@seberg Cool. Let me know if there's anything I can do to help.

seberg commented:
We may have to rebase now, since I removed the deprecation in gh-22004 so this should be an error now.

nschloe commented:
I manually rebased.

nschloe committed:
deprecate scalar conversion for arrays of ndim > 0

nschloe committed:
deprecate scalar conversion for arrays of ndim > 0

nschloe committed:
deprecation test fix

nschloe committed:
deprecation test fix

nschloe committed:
fix code block

nschloe committed:
fix code block

nschloe commented:
@seberg Let me know if there's anything I can do to advance this.

nschloe committed:
Merge branch 'bug-10404' of github.com:nschloe/numpy into bug-10404

nschloe committed:
Merge branch 'main' into bug-10404

nschloe committed:
update dates

nschloe committed:
lint fix

nschloe commented:
I've updated SciPy again to be warnings-free after this PR (https://github.com/scipy/scipy/pull/17698). Changes are now usually minimal.

I'd be great if this PR could be looked at again. We'll soon be celebrating it's 5th birthday. :)

asmeurer commented:
Are there plans to also deprecate the related reverse broadcasting behavior which was discussed on the list?

mattip commented:
I marked this for triage review, it also has a conflict.

seberg committed:
Merge branch 'main' into bug-10404

seberg commented:
Planning to put this in later today and give it a shot.  There is always the (correct) worry here that it might be noisy for downstream.  OTOH, this has also repeatedly come up over the time.

We had brought it up very briefly yesterday, but @mattip, if you had a different intention when marking it, just let me know.

What makes me optimistic is that the scipy changes are far larger than the other libraries (all of which seem very simple, the optimizers really were most of it probably).

So to be clear:  I am happy to roll it back if this annoys downstream libraries or anyone has serious concerns (libs like dask or xarray for example).
As for downstream users, I don't mind taking this deprecation slow.

mroeschke commented:
@seberg in pandas, how should we treat size 1 arrays with PyObject list/tuples that we want to treat as scalars for assignment e.g.

```
-> values[indexer] = casted
(Pdb) values
array([10, 11, 12], dtype=object)
(Pdb) indexer
array([1])
(Pdb) casted
array([(61,)], dtype=object)
(Pdb) n
(Pdb) values
array([10, (61,), 12], dtype=object)  # want to assign a tuple
(Pdb) np.__version__
'1.24.2'

---

(Pdb) values[indexer] = casted[0]  # correct workaround for the deprecation?
(Pdb) values
array([10, 61, 12], dtype=object)  # 61 is assigned instead of (61,)
```

asmeurer commented:
@mroeschke you can work around that by doing `values[indexer] = casted[0, ...]`. That forces the right-hand side to be a 0-D array instead of a scalar, which prevents NumPy from trying to call asarray on it.

asmeurer commented:
Even this deprecation notwithstanding it's a good idea to use 0-D arrays instead of scalars everywhere if you're dealing with object dtype arrays with elements that themselves could be interpreted as arrays by `asarray`. 

The issue is that `casted[0]` is literally type `tuple` (i.e., not a NumPy type anymore), so there's inherent ambiguity whether `(61,)` should be interpreted as the tuple object or as the 1-D array `array([61])` in pretty much any context where NumPy could call `asarray` (which is virtually everywhere).

neutrinoceros commented:
Provided my reading of the patch, and in particular the comment attached to the deprecation code, is correct, the error that will be raised after this deprecation is expired is `TypeError` ?

For context, I'm working on a case where the implicit conversion from size-1 array to scalar happens in a `try: ... except TypeError:` block, and landing in the `except` branch would be acceptable.

seberg commented:
Yes, it will be a `TypeError`.
